### ISSUE TITLE
urlEncodeMap don't right resolve null value

### ISSUE THREAD

**Issue by fuyong859050943** [flutterchina/dio#422]
in query or formdata
```json
{
  "username": "test",
  "password": null
}
```
it will generate like `username=test&password=` not like `username=test&password=null`
`Transformer.urlEncodeMap` also has the same problem?
i have changed like this:
```dart
  /// Deep encode the [Map<String, dynamic>] to percent-encoding.
  /// It is mostly used with  the "application/x-www-form-urlencoded" content-type.
  static String urlEncodeMap(data) {
    StringBuffer urlData = new StringBuffer("");
    bool first = true;
    void urlEncode(dynamic sub, String path) {
      if (sub is List) {
        for (int i = 0; i < sub.length; i++) {
          urlEncode(sub[i],
              "$path%5B${(sub[i] is Map || sub[i] is List) ? i : ''}%5D");
        }
      } else if (sub is Map) {
        sub.forEach((k, v) {
          if (path == "") {
            urlEncode(v, "${Uri.encodeQueryComponent(k)}");
          } else {
            urlEncode(v, "$path%5B${Uri.encodeQueryComponent(k)}%5D");
          }
        });
      } else {
        if (!first) {
          urlData.write("&");
        }
        first = false;
        // when value is null replace
        urlData.write("$path=${sub == null ? '' : Uri.encodeQueryComponent(sub.toString())}");
      }
    }

    urlEncode(data, "");
    return urlData.toString();
  }
```

**Comment by amrosama2020** [comment]
I don't think this is an issue, null should be encoded to empty string unless it is a string with the value "null".

If you really need to send the word null I suggest the following while building the request body:
```dart
{
  "username": "test",
  "password": password??"null"
}
```

**Comment by wendux** [comment]
Fixed at v2.1.14.

**Comment by fuyong859050943** [comment]
> I don't think this is an issue, null should be encoded to empty string unless it is a string with the value "null".
> 
> If you really need to send the word null I suggest the following while building the request body:
> 
> ```dart
> {
>   "username": "test",
>   "password": password??"null"
> }
> ```

Can you read my question clearly and answer it?
